src/Surfnet/StepupMiddleware/MiddlewareBundle/Console/Command/BootstrapIdentityWithSmsSecondFactorCommand.php

```diff
@@ -0,0 +1,220 @@
+            ->addArgument('preferred-locale', InputArgument::REQUIRED, 'The preferred locale of the identity to create')
+            ->addArgument(
+                'phone-number',
+                InputArgument::REQUIRED,
+                'The phone number of the user should be formatted like "+31 (0) 612345678"'
+            )
+            ->addArgument(
+                'registration-status',
+                InputArgument::REQUIRED,
+                'Valid arguments: unverified, verified, vetted'
+            );
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        /** @var Container $container */
+        $container = $this->getApplication()->getKernel()->getContainer();
+        $tokenStorage = $container->get('security.token_storage');
+        $identityRepository = $container->get('surfnet_stepup_middleware_api.repository.identity');
+        $unverifiedSecondFactorRepository = $container->get(
+            'surfnet_stepup_middleware_api.repository.unverified_second_factor'
+        );
+        $verifiedSecondFactorRepository = $container->get(
+            'surfnet_stepup_middleware_api.repository.verified_second_factor'
+        );
+        $pipeline = $container->get('surfnet_stepup_middleware_command_handling.pipeline.transaction_aware_pipeline');
+        $eventBus = $container->get('surfnet_stepup_middleware_command_handling.event_bus.buffered');
+        $connection = $container->get('surfnet_stepup_middleware_middleware.dbal_connection_helper');
+
+        $tokenStorage->setToken(
+            new AnonymousToken('cli.bootstrap-identity-with-sms-token', 'cli', ['ROLE_SS', 'ROLE_RA'])
+        );
+
+        $nameId = new NameId($input->getArgument('name-id'));
+        $institution = new Institution($input->getArgument('institution'));
+        $commonName = $input->getArgument('common-name');
+        $email = $input->getArgument('email');
+        $preferredLocale = $input->getArgument('preferred-locale');
+        $registrationStatus = $input->getArgument('registration-status');
+        $phoneNumber = $input->getArgument('phone-number');
+        $identity = false;
+
+        $output->writeln(
+            sprintf(
+                '<notice>Adding a %s SMS token for %s</notice>',
+                $registrationStatus,
+                $commonName
+            )
+        );
+
+        if ($identityRepository->hasIdentityWithNameIdAndInstitution($nameId, $institution)) {
+            $output->writeln(
+                sprintf(
+                    '<notice>An identity with name ID "%s" from institution "%s" already exists, using that identity</notice>',
+                    $nameId->getNameId(),
+                    $institution->getInstitution()
+                )
+            );
+            $identity = $identityRepository->findOneByNameIdAndInstitution($nameId, $institution);
+        }
+
+        $connection->beginTransaction();
```

Comment:
Why are you using nested transactions when using the transaction aware pipeline. Are  you sure this is safe to use?

Reply:
My work is based on the tried and tested yubikey bootstrap script. I do not see nested transactions at work here. At least not according to this definition:

https://www.doctrine-project.org/projects/doctrine-dbal/en/2.10/reference/transactions.html#transaction-nesting

Can you elaborate?

> Are you sure this is safe to use?

The tests I performed worked out fine, whenever illegal operations where applied, the rollback feature worked a charm.

Reply:
In the TransactionAwarePipeline multiple connections are used and separate transactions are used. So you need to be sure that if a rollback is performed this will happening on all connections. Personally I think it's not safe to assume the old code is very well battle tested based on earlier discovered bugs.